### DISCUSSION TITLE
Croped and replaced Image Retuns Black

### DISCUSSION THREAD

**Discussion by daneshgit** [pymupdf/PyMuPDF#1216] [Looking for help]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)
 I am applying redaction on the following pdf .In that pdf the image area is blacked out during the redaction process, as an alternate I am trying with  the pixmap the get the original image without blacked out and replaced it after the redaction process on that particular image. Could be helpful is there is any alternate solution for removing the blacked 
[combined-pdf-92.pdf](https://github.com/pymupdf/PyMuPDF/files/7007645/combined-pdf-92.pdf)
out image.

## To Reproduce (mandatory)
I given the following apply_redactions()   , apply_redactions(images=1) , apply_redactions(images=2)
Following is the code I used:
basepix = page.getPixmap(clip=fitz.Rect(image['bbox']),alpha=False)
page.addRedactAnnot(image['bbox'], fill=colors.to_rgb("white"), text=" ")
page.apply_redactions()
page.insert_image(image['bbox'],pixmap=basepix,overlay=True,keep_proportion=True)

## Expected behavior (optional)
I expect the same image original image.

## Screenshots (optional)
Output Image
![image](https://user-images.githubusercontent.com/26829224/129906492-d36f7b33-ed3e-42d8-8a80-541d9910ce17.png)

Input Image
![image](https://user-images.githubusercontent.com/26829224/129906885-62fbe6e2-e8d0-4c4b-878a-8a4be0c0807a.png)

Two images overwritten example
![image](https://user-images.githubusercontent.com/26829224/129909171-29570a1b-7159-4fc2-b154-f0da7209a4ae.png)

## Your configuration (mandatory)
 - Operating system, potentially version and bitness
 - Python version, bitness
 - PyMuPDF version, installation method (**wheel** or **generated** from source).

For example, the output of `print(sys.version, "\n", sys.platform, "\n", fitz.__doc__)` would be sufficient (for the first two bullets).

**Comment by JorjMcKie** [comment]
This a known upstream bug:
If you choose to whiten out areas of images which overlap redaction rectangles, then this will lead to undesired results if the image has a mask (i.e. an alpha channel). This is documented in the manual!

You have 2 options
1. do not touch overlapping images at all: `fitz.PDF_REDACT_IMAGE_NONE`
2. remove overlapping images completely: `fitz.PDF_REDACT_IMAGE_REMOVE`

So instead of your workaround just choose option 1 and do not let the method default to `fitz.PDF_REDACT_IMAGE_PIXELS`.
If you abslutely need to see whitened subareas of images that overlap, you must simulate the effect for alpha images by overlaying it with a redaction rectangle with a white background (use fill parameter).

**Comment by JorjMcKie** [comment]
Sorry, did not read your issue completely:
I tested your PDF by removing the word "Phone" using this snippet
```python
rl = page.search_for("phone")
for r in rl:
    page.add_redact_annot(r)
page.apply_redactions(images=fitz.PDF_REDACT_IMAGES_NONE)
```
and so no black images at all.

**Comment by daneshgit** [comment]
Hi 'phone' is not an image  hence its redacting.
I tried with "067024" which is located in the page end , the result is its not redacting using the below method.

rl = page.search_for("067024")
            for r in rl:
                page.add_redact_annot(r,fill=colors.to_rgb('red'))
                page.apply_redactions(images=0)

As an alternative I got the coordinates of "067024" and redacted using the following method

1.This returns black image

            clip = fitz.Quad(((216,813), (257,813), (257,822), (216,822)))
            page.add_redact_annot(clip,fill=colors.to_rgb('red'))
            page.apply_redactions()

2.This returns like applying just a rectangle in the foreground but if we see  in the background by editing pdf the keyword remains

            clip = fitz.Quad(((216,813), (257,813), (257,822), (216,822)))
            page.add_redact_annot(clip,fill=colors.to_rgb('red'))
            page.apply_redactions(images=0)
![MicrosoftTeams-image](https://user-images.githubusercontent.com/26829224/130020586-c0bd6878-b588-4e0b-9b04-c357a00adae9.png)

3.This returns like removing all images related to "067024"
 	    clip = fitz.Quad(((216,813), (257,813), (257,822), (216,822)))
            page.add_redact_annot(clip,fill=colors.to_rgb('red'))
            page.apply_redactions(images=1)

4.This returns the Black image
	    clip = fitz.Quad(((216,813), (257,813), (
                    257,822), (216,822)))
            page.add_redact_annot(clip,fill=colors.to_rgb('red'))
            page.apply_redactions(images=2)



Can I have the solution to remove "067024" without affecting anything others

**Comment by JorjMcKie** [comment]
> Can I have the solution to remove "067024" without affecting anything others

No, not "<0677024>", because this is **_part_** of an image. You won't find string "<0677024>".
String "0677024" does get removed fine (inside the box "Office use only"). That text has been placed above the image and can therefore be redacted.

**Comment by daneshgit** [comment]
Hi , As Its is in the document apply_redaction(images=2)  removes only the overlapping  part of the image on which redaction is annotated. May I know why its blacked out all.

Input image with the following code:
clip = fitz.Quad(((216,813), (257,813), (257,822), (216,822)))
page.add_redact_annot(clip,fill=colors.to_rgb('white'))

[test18.pdf](https://github.com/pymupdf/PyMuPDF/files/7026562/test18.pdf)

Result after applying the following code
apply_redaction(images=2)
[test20.pdf](https://github.com/pymupdf/PyMuPDF/files/7026565/test20.pdf)

**Comment by JorjMcKie** [comment]
It also is documented, that this part of MuPDF code is not bug free. For images with a mask (transparency present), and also some black and white images, that logic is not (always) working.
They know about the bug.

**Comment by agsdineshDev** [comment]
Whoever fixes the transparent image apply reaction issues and the layered image issue described above will be reward with bitcoin worth 1000 Australian Dollar. 
Share the fixed python code and bitcoin address with us once upon verification. The reward will be processed. 

![image](https://user-images.githubusercontent.com/63396565/131656933-aea0ebaa-5f60-417b-84eb-57bf59a05620.png)
Bag your reward.

Good luck!!!

**Reply by JorjMcKie** [reply]
Interesting offer!
I have developed something which results in this for your example:
[curate-images.pdf](https://github.com/pymupdf/PyMuPDF/files/7093353/curate-images.pdf)
I do not want to attach my code in this thread: if you are serious, then I am serious as well.
Please let me know how to proceed further 😎
